playground/frontend/lib/pages/playground/states/playground_state.dart

```diff
@@ -45,95 +49,142 @@ const kCachedResultsLog =
+          one: descriptor,
+        );
+      }
+    }
+
+    return result;
+  }
+
+  // TODO: Return full, then shorten.
   String get examplesTitle {
-    final name = _selectedExample?.name ?? kTitle;
+    final name = snippetEditingController?.selectedExample?.name ?? kTitle;
     return name.substring(0, min(kTitleLength, name.length));
   }
 
-  ExampleModel? get selectedExample => _selectedExample;
+  ExampleModel? get selectedExample =>
+      snippetEditingController?.selectedExample;
+
+  SDK? get sdk => _sdk;
+
+  SnippetEditingController? get snippetEditingController =>
+      _snippetEditingControllers[_sdk];
+
+  SnippetEditingController requireSnippetEditingController() {
+    final controller = snippetEditingController;
+
+    if (controller == null) {
+      throw Exception('SDK is not set.');
+    }
 
-  SDK get sdk => _sdk;
+    return controller;
+  }
 
-  String get source => codeController.rawText;
+  String? get source => snippetEditingController?.codeController.text;
 
   bool get isCodeRunning => !(result?.isFinished ?? true);
 
   RunCodeResult? get result => _result;
 
-  String get pipelineOptions => _pipelineOptions;
+  String? get pipelineOptions => snippetEditingController?.pipelineOptions;
 
   Stream<int>? get executionTime => _executionTime?.stream;
 
   bool get isExampleChanged {
-    return selectedExample?.source != source || _arePipelineOptionsChanges;
-  }
-
-  bool get _arePipelineOptionsChanges {
-    return pipelineOptions != (_selectedExample?.pipelineOptions ?? '');
+    return snippetEditingController?.isChanged ?? false;
   }
 
   bool get graphAvailable =>
       selectedExample?.type != ExampleType.test &&
       [SDK.java, SDK.python].contains(sdk);
 
-  void setExample(ExampleModel example) {
-    _selectedExample = example;
-    setSdk(example.sdk, notify: false);
-    _pipelineOptions = example.pipelineOptions ?? '';
-    codeController.text = example.source ?? '';
+  void setExample(
+    ExampleModel example, {
+    required bool setCurrentSdk,
+  }) {
+    if (setCurrentSdk) {
+      setSdk(example.sdk, loadDefaultIfNot: false, notify: false);
+      snippetEditingController!.selectedExample = example;
+    } else {
+      final controller = _getOrCreateSnippetEditingController(
+        example.sdk,
+        loadDefaultIfNot: false,
+      );
+      controller.selectedExample = example;
+    }
+
     _result = null;
     _executionTime = null;
     setOutputResult('');
     notifyListeners();
   }
 
-  void setSdk(SDK sdk, {bool notify = true}) {
+  void setSdk(
+    SDK sdk, {
+    required bool loadDefaultIfNot,
+    bool notify = true,
+  }) {
     _sdk = sdk;
-    codeController.language = sdk.highlightMode;
+    _getOrCreateSnippetEditingController(sdk,
+        loadDefaultIfNot: loadDefaultIfNot);
 
     if (notify) {
       notifyListeners();
     }
   }
 
-  set source(String source) {
-    codeController.text = source;
+  void setSource(String source) {
```

Comment:
We recently turned this from `setSource` to `set source` following the linter, but now we need nullable getter with a non-nullable setter, and that does not work, so back to the method again.